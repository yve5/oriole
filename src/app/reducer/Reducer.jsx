import { CVSS4_CHANGE_VALUE, CVSS4_RESET_SCORE } from '../resources/constants';
import { getDefaultMetrics } from '../../cvss4';

const initialState = {
  metrics: getDefaultMetrics(),
};

const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case CVSS4_RESET_SCORE:
      return {
        ...state,
        ...action,
      };

    case CVSS4_CHANGE_VALUE: {
      const updatedMetric = {};
      updatedMetric[action.metric] = action.value;

      const copyAction = { ...action };
      delete copyAction.metric;
      delete copyAction.value;

      return {
        ...state,
        ...copyAction,
        metrics: {
          ...state.metrics,
          ...updatedMetric,
        },
      };
    }

    default:
      return state;
  }
};

export default reducer;
