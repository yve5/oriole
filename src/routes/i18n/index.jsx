import I18nPage from './components/I18n';

export default I18nPage;
