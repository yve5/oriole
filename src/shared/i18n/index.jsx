import LangSelect from './components/LangSelect';
import LangSwitch from './components/LangSwitch';
import I18n from './components/I18n';
import i18nReducer from './reducer';

export { I18n, LangSelect, LangSwitch, i18nReducer };
