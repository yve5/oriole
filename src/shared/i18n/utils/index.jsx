import { checkTermOrder, convertCollectionToArray } from './Term';
import { getInitialState } from './InitialState';
import { checkMissingTerms } from './Missing';
import { formatDate } from './Format';

export {
  formatDate,
  checkTermOrder,
  getInitialState,
  checkMissingTerms,
  convertCollectionToArray,
};
