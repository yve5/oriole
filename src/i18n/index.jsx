import fr from './fr';
import en from './en';
import { tsl } from '../shared/i18n/utils/Translate';

const i18n = {
  tsl: (term, args) => tsl(fr, term, args),
  dictionaries: [fr, en],
  lang: fr.LANG,
};

export default i18n;
