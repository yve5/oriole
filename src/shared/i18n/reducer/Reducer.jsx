import en from '../dictionaries/en';
import fr from '../dictionaries/fr';

import { tsl } from '../utils/Translate';
import {
  I18N_CHANGE_LANG,
  I18N_SWITCH_LANG,
  I18N_LOCAL_USER_LANG,
} from '../resources/constants';

const initialState = {
  tsl: (term, args) => tsl(en, term, args),
  dictionaries: [en, fr],
  lang: en.LANG,
};

const reducer = (dictionaries, localStorageName = I18N_LOCAL_USER_LANG) => {
  const getIndex = (dicos, lang) => {
    const index = dicos.findIndex(({ LANG }) => LANG === lang);
    return index === -1 ? 0 : index;
  };

  let innerState = { ...initialState };

  if (Array.isArray(dictionaries) && dictionaries[0]) {
    const subDicos = [];
    dictionaries.forEach((dico) => {
      if (dico.LANG === en.LANG) {
        subDicos.push({ ...en, ...dico });
      } else if (dico.LANG === fr.LANG) {
        subDicos.push({ ...fr, ...dico });
      } else {
        subDicos.push(dico);
      }
    });

    innerState = {
      ...innerState,
      dictionaries: subDicos,
      lang: subDicos[0].LANG,
      tsl: (term, args) => tsl(subDicos[0], term, args),
    };
  }

  const localLang = localStorage.getItem(localStorageName);

  if (localLang) {
    const subDicos = innerState.dictionaries;
    const initialIndex = getIndex(subDicos, localLang);

    innerState = {
      ...innerState,
      lang: localLang,
      tsl: (term, args) => tsl(subDicos[initialIndex], term, args),
    };
  }

  return (state = innerState, action = {}) => {
    switch (action.type) {
      case I18N_CHANGE_LANG: {
        const currentIndex = getIndex(state.dictionaries, action.lang);

        return {
          ...state,
          ...action,
          tsl: (term, args) =>
            tsl(state.dictionaries[currentIndex], term, args),
        };
      }

      case I18N_SWITCH_LANG: {
        const currentIndex = getIndex(state.dictionaries, state.lang);

        const nextIndex = (currentIndex + 1) % state.dictionaries.length;
        const nextDico = state.dictionaries[nextIndex];
        const nextLang = nextDico.LANG;

        localStorage.setItem(localStorageName, nextLang);

        return {
          ...state,
          ...action,
          lang: nextLang,
          tsl: (term, args) => tsl(state.dictionaries[nextIndex], term, args),
        };
      }

      default:
        return state;
    }
  };
};

export default reducer;
