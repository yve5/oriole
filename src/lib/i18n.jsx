import { I18n, LangSelect, LangSwitch, i18nReducer } from '../shared/i18n';
import {
  trim,
  formatDate,
  checkTermOrder,
  checkMissingTerms,
  convertCollection,
  checkLocalTermOrder,
  convertLocalCollection,
} from '../shared/i18n/utils';

export {
  I18n,
  trim,
  formatDate,
  LangSelect,
  LangSwitch,
  i18nReducer,
  checkTermOrder,
  checkMissingTerms,
  convertCollection,
  checkLocalTermOrder,
  convertLocalCollection,
};
