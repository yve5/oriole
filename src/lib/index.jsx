import Theme from '../shared/theme';
import { I18n, i18nState, i18nReducer } from '../shared/i18n';

export { Theme, I18n, i18nState, i18nReducer };
