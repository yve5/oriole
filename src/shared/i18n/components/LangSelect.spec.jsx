import React from 'react';
import configureMockStore from 'redux-mock-store';

import { thunk } from 'redux-thunk';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';

import reducer from '../reducer';
import { LangSelect } from '../../../lib/i18n';

const mockStore = configureMockStore([thunk]);
const defaultStore = { i18n: reducer()() };

describe('LangSelect', () => {
  it('should match expected snapshot. default', async () => {
    const { asFragment, getByTestId } = render(
      <Provider store={mockStore(defaultStore)}>
        <LangSelect />
      </Provider>
    );

    const selectElement = getByTestId('i18n-lang-select');
    fireEvent.change(selectElement, { target: { value: 'fr' } });

    expect(selectElement.value).toBe('fr');
    expect(asFragment()).toMatchSnapshot();
  });
});
