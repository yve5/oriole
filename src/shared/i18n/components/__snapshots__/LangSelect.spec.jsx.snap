// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LangSelect > should match expected snapshot. default 1`] = `
<DocumentFragment>
  <select
    class="form-select"
    data-testid="i18n-lang-select"
  >
    <option
      value="en"
    >
      English
    </option>
    <option
      value="fr"
    >
      Fran√ßais
    </option>
  </select>
</DocumentFragment>
`;
